### PR TITLE
basic aldex plugin

### PR DESCRIPTION
**Pull request by dgiguer**
- visualize differentially expressed features (effect plot, volcano plot, MA plot)
- extract differentially expressed features

### CODE REVIEW

**Comment by mortonjt** [comment]
Hi Daniel, this is awesome! Let me give it quick test

File: q2_aldex2/_visualizer.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        img_effect = os.path.join(output_dir, "effect_q_plot.png")
+
+        plt.savefig(img_effect)
+
+        # get path of index file
+        index = os.path.join(TEMPLATES, 'index.html')
+
+        # plot_name displays whether effect or MA
+        q2templates.render(index, output_dir, context={'plot_name': type})
```
Comment (mortonjt) [inline q2_aldex2/_visualizer.py]: this doesn't look right ... `type` is a reserved keyword.   Are you sure you didn't mean to put in `test`?

File: q2_aldex2/_method.py [hunk 2]
```diff
@@ -48,6 +61,32 @@ def aldex2(table: pd.DataFrame,
...
+    # absolute values needed for effect or difference to see change in either
+    # condition
+    if effect_threshold > abs(table['effect']).max():
+        raise ValueError("You have selected an effect threshold that exceeds maximum effect size (-p--effect-threshold). Choose a lower threshold, or be aware that there there will be no features in the output.")
+
+    if difference_threshold > abs(table['diff.btw']).max():
+        raise ValueError("You have selected a difference threshold that exceeds maximum difference (-p--difference-threshold). Choose a lower threshold, or be aware that there will be no features in the output.")
+
+    # subset the table if it psases all the threshold
+    differentials_sig = table[(table[effect_statistic_function] <= sig_threshold) & (abs(table['effect']) > effect_threshold) & (abs(table['diff.btw']) > difference_threshold)]
```
Comment (mortonjt) [inline q2_aldex2/_method.py]: Totally optional - but a good python practice is to have everything within 80 characters.  
So lines like the above line will be split across 2-3 lines.  It is easier for future developers to read the code.

But I'll leave it up to you if you want to do this or not.

File: tutorial.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+# p is effect plot or MA plot. currently only effect plot is available.
+qiime aldex2 effect-plot --i-table aldex_output/differentials.qza --p-type effect --o-visualization selex_effect
+
+# visualize using the qiime tools view
+# this will pop up the visualizer in a chrome window
+qiime tools view selex_effect.qzv
+
+```
+
+This should generate several effect plots. To extract the differentially abundant points, use `extract-differences` and look at the `differentials` file by extracting it
```
Comment (mortonjt) [inline tutorial.md]: `points` -> `microbes`

File: tutorial.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+Install qiime2 and activate your conda environment ([see qiime2 Getting Start](https://docs.qiime2.org/2019.7/getting-started/)).
+
+Install this github repo:
+```
+# clone to working directory for qiime
+git clone https://github.com/dgiguer/q2-aldex2 ./q2-aldex2
+cd ./q2-aldex2
+pip install -e .
+cd ../
```
Comment (mortonjt) [inline tutorial.md]: A much easier way to install would be

```
pip install git+https://github.com/dgiguer/q2-aldex2
```

File: tutorial.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+```
+qiime tools export --input-path gut-test/sig_gut.qza --output-path differentially-expressed-features
+```
+
+# Tutorial with raw count tables
+
+This tutorial is meant for users starting from scratch.
+
+Install qiime2 and activate your conda environment ([see qiime2 Getting Start](https://docs.qiime2.org/2019.7/getting-started/)).
```
Comment (mortonjt) [inline tutorial.md]: qiime2 Getting Start -> qiime2 Getting Started

File: tutorial.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+```
+qiime feature-table filter-samples \
+  --i-table table.qza \
+  --m-metadata-file sample-metadata.tsv \
+  --p-where "[body-site]='gut'" \
+  --o-filtered-table gut-table.qza
+```
+
+The next step will be to run ALDEx2. The full pipeline is implemented in the `aldex2` function The input is the `FeatureTable[Frequency]`, as well as a metadata file. The metadata file is necessary for defining the different groups you will be testing. For this tutorial, the groups are identified by the subject column from the metadata file. ALDEx2 automatically adds a pseudocount to remove zeros from the data, and filters any samples with 0 reads.
```
Comment (mortonjt) [inline tutorial.md]: Maybe also worth mentioning that the pseudocount is a Dirichlet prior with the assumption that all microbes are present (it provides some more motivation about why pseudocounts are ok under certain conditions).

File: tutorial.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+```
+qiime aldex2 aldex2 --i-table gut-table.qza --m-metadata-file gut-metadata.tsv --p-condition subject --output-dir gut-test
```
Comment (mortonjt) [inline tutorial.md]: I'd recommend changing gut-metadata.tsv to sample-metadata.tsv so that it is consistent with the moving pictures tutorial.

**Review by mortonjt — APPROVED** [review]
Hi @dgiguer , this is awesome!

I've added a couple of comments - totally up to you if you want to address them or not.
Once you are happy with the final revision, I'll transfer ownership of this repository to you, so that you'll have full control.

**Comment by fedarko** [comment]
Hi @dgiguer! This is looking really exciting, thanks to you and @mortonjt for your work on making these methods accessible through QIIME 2!

Going through the current version of the tutorial, I ran into a few problems. Here's a list of the problems I ran into, and some potential solutions to make these easier for future users.

## Error 1: `/assets/` aren't installed

I installed the q2-aldex2 plugin from your repository (via the command `pip install git+https://github.com/dgiguer/q2-aldex2`, as @mortonjt suggested above), but it looks like the `assets/` folder wasn't installed along with q2-aldex2. This causes `qiime aldex2 effect-plot` to fail (saying that `.../assets/index.html` wasn't available). 

I'm guessing that when you run `pip install -e .` after cloning ALDEx2's code, as you currently suggest in your docs, the `assets/` folder is accessible since it's already on your computer. Whereas, if you install things like I installed, pip ignores `assets/` since it doesn't look like it contains any python code.

I guess the problem here is mostly with me... but if you want to release q2-aldex2 on something like PyPI, this will probably need to be addressed.

### Long-term solution
You could probably fix this by mentioning the `assets/` in `package_data` in your setup.py -- see, for example, [these docs](https://docs.python.org/3/distutils/setupscript.html#installing-package-data). This should ensure that `assets/` are installed regardless of how pip installs q2-aldex2.

### Short-term solution
Anyway, I was able to solve this problem for my particular system by just navigating to where I had q2-aldex2 installed (for me, `/anaconda3/envs/qiime2-2019.7/lib/python3.6/site-packages/q2_aldex2/`), creating an `assets/` directory, and downloading the `index.html` file you have in your repository. This let me create an effect plot!

## Error 2: The images in the effect plot hang off the bottom of the screen, without being scrollable

Here's a screenshot of what this looks like:
Screen Shot 2019-08-08 at 9 40 45 PM

### Solution
If you remove the block of CSS that uses `overflow: hidden;` ([the three lines starting here](https://github.com/dgiguer/q2-aldex2/blob/master/q2_aldex2/assets/index.html#L8) in your index.html file), this adds back a scrollbar and makes these plots a lot more viewable :)

## Error 3: Feature names in the differentials sometimes start with `X`?

I ran into another error while trying something else (sorry). Here's where things get confusing for me.

I've written [another software package](https://github.com/biocore/qurro) that can visualize the feature differentials produced by tools like ALDEx2 or [Songbird](https://github.com/biocore/songbird), so one of the first things I wanted to try was throwing the ALDEx2 differentials into that to display them. But it looks like some of the feature names in the differentials start with `X`, for some reason:

```bash
featureid	rab.all	rab.win.subject.1	rab.win.subject.2	diff.btw	diff.win	effect	overlap	we.ep	we.eBH	wi.ep	wi.eBH
#q2:types	numeric	numeric	numeric	numeric	numeric	numeric	numeric	numeric	numeric	numeric	numeric
X4b5eeb300368260019c1fbc7a3c718fc	6.63145167899188	11.6731196654531	-2.37787617761982	-14.1583926895949	2.7920342175922	-5.28275450057488	0.00027407495720991	0.00263067309096698	0.0590136803348092	0.0285714285714286	0.198356473713876
fe30ff0f71a38a39cf1717ec2be3a2fc	-2.00086318269129	-1.19884226951553	-2.49373890849691	-1.78111213948672	4.70104391436084	-0.307961088355758	0.35797684861086	0.45632650903061	0.683613347117657	0.525446428571429	0.718806648436809
X868528ca947bc57b69ffdf83e6b73bae	6.37505975414147	-1.74944651319553	10.4532747351492	12.2375842992755	2.37908035878766	5.26247339611525	0.00027407495720991	0.00297567857690031	0.0605968456054936	0.0285714285714286	0.198356473713876
```

When I look through the feature IDs in the corresponding BIOM table, I don't see `X4b5eeb300368260019c1fbc7a3c718fc` anywhere. But `4b5eeb300368260019c1fbc7a3c718fc` is a feature ID in this table. Strangely enough, `fe30ff0f71a38a39cf1717ec2be3a2fc` is unaffected.

I'm not sure if this is intentional or not, but it messes up downstream visualization of these differentials (at least for me :), but I imagine it will affect other tools also). Is there any reason certain features have their IDs prefixed with `X` in this output? I'm not sure where this bug comes from (or if it even is a bug). @mortonjt have you run into this sort of thing before?

## In any case
I know this was a lot of stuff (sorry to inundate you with comments), but hopefully it all helps make this tool more widely useful to people! I'm looking forward to using ALDEx2 more.

File: q2_aldex2/_visualizer.py (hunk 1 above)
Comment (dgiguer) [reply q2_aldex2/_visualizer.py]: yup, good catch! updated.

File: q2_aldex2/_method.py (hunk 2 above)
Comment (dgiguer) [reply q2_aldex2/_method.py]: fixed!

File: tutorial.md (hunk 3 above)
Comment (dgiguer) [reply tutorial.md]: updated to features

File: tutorial.md (hunk 4 above)
Comment (dgiguer) [reply tutorial.md]: agreed!

File: tutorial.md (hunk 5 above)
Comment (dgiguer) [reply tutorial.md]: good catch!

File: tutorial.md (hunk 6 above)
Comment (dgiguer) [reply tutorial.md]: updated, with a link to the paper for more information

File: tutorial.md (hunk 7 above)
Comment (dgiguer) [reply tutorial.md]: agreed!

**Comment by dgiguer** [comment]
I pushed commits to address these comments! thanks for taking the time to review this @mortonjt @fedarko !

**Comment by fedarko** [comment]
@dgiguer Thanks so much, this is amazing! I'm happy that I can start using ALDEx2 through QIIME 2 now :)

**Comment by mortonjt** [comment]
@fedarko nice catch with the X!  That is the default for R.

@dgiguer merging this in now and will transfer over.  Thanks!

**Pull request merged**